File: Python/ScientificCalculator/functions/first_degree_equation.py
"""
first_degree_equation.py

Solves single-variable equations.

Supported:
- Exactly one variable: a
- Exactly one '='
- +  -  *  /
- Parentheses
- Decimals
- Fractions

Examples:

a + 4 = 6
2a + 5 = 17
5a - 3 = 2a + 9
a/5 = 8
5/a = 20
1/(a+2)=3
"""

from decimal import Decimal


def solve(expression):
    """
    Solve an equation in one variable 'a'.

    Returns:
        Decimal solution.
    """

    if expression.count("=") != 1:
        raise ValueError(
            "Equation must contain exactly one '='."
        )

    left, right = expression.split("=")

    def f(a):

        scope = {
            "__builtins__": None,
            "a": Decimal(str(a))
        }

        return eval(left, scope) - eval(right, scope)

    # -----------------------------
    # Search for a sign change
    # -----------------------------

    x = Decimal("-100000")
    step = Decimal("1")

    previous = f(x)

    while x <= Decimal("100000"):

        x2 = x + step

        try:
            current = f(x2)
        except Exception:
            x = x2
            previous = None
            continue

        if previous is not None:

            if current == 0:
                return x2

            if previous == 0:
                return x

            if previous * current < 0:

                low = x
                high = x2

                # Binary search

                for _ in range(120):

                    mid = (low + high) / 2

                    value = f(mid)

                    if abs(value) < Decimal("1e-40"):
                        return mid

                    if f(low) * value < 0:
                        high = mid
                    else:
                        low = mid

                return (low + high) / 2

        previous = current
        x = x2

    raise ValueError("No real solution found.")

File: Java/ScientificCalculator/src/functions/FirstDegreeEquation.java
package functions;

import java.util.List;

import parser.EquationTokenizer;
import parser.EquationParser;
import parser.LinearExpression;


/*
 * FirstDegreeEquation
 *
 * Solves equations of the form:
 *
 * ax + b = cx + d
 *
 * Examples:
 *
 * 2x+4=10
 * 3x-5=2x+9
 * 4(x-3)+5=2(x+1)
 *
 */

public class FirstDegreeEquation {


    // ==========================
    // Solve equation
    // ==========================

    public static double solve(
            String equation
    ) {


        String[] sides =
                equation.split("=");


        if (sides.length != 2) {

            throw new IllegalArgumentException(
                "Equation must contain one = sign"
            );

        }



        LinearExpression left =
                parseSide(
                    sides[0]
                );


        LinearExpression right =
                parseSide(
                    sides[1]
                );



        // Move everything left:
        //
        // ax+b - (cx+d)=0

        double coefficient =
                left.getCoefficient()
                -
                right.getCoefficient();


        double constant =
                left.getConstant()
                -
                right.getConstant();



        if (coefficient == 0) {


            if (constant == 0) {

                throw new ArithmeticException(
                    "Infinite solutions"
                );

            }


            throw new ArithmeticException(
                "No solution"
            );

        }



        return -constant / coefficient;

    }



    // ==========================
    // Parse one side
    // ==========================

    private static LinearExpression parseSide(
            String side
    ) {


        List<String> tokens =
                EquationTokenizer.tokenize(
                        side
                );


        EquationParser parser =
                new EquationParser(
                        tokens
                );


        return parser.parse();

    }

}
